### PR TITLE
fix: acf fields not resolving with preview

### LINKED ISSUE #154: ACF fields cannot be queried in previews
### Description

It looks like ACF fields cannot be queried in previews (at least not the preview data).
My Goal: Be able to access ACF preview data from the API.

This is my query:
```
query MyQuery2 {
  post(id: 125, idType: DATABASE_ID, asPreview: true) {
    title
    skeleton {
      textField
      skeleton {
        ... on SkeletonSkeletonHeadingCopyLayout {
          title
        }
      }
    }
  }
}
```

The post's title gets the preview value correctly but both ACF fields are returned with their published values (not the preview).

I've also already tried many variations of this query using revisionOf, or with asPreview: false and accessing by {preview {...}}, or even querying using nodeByUri. Never works.

Screenshot 2024-01-05 at 8 52 22 am (1)

### Steps to reproduce

1. I am using WordPress 6.4.2 currently on a localhost from Local by Flywheel PHP version 8.2.8.
2. Deactivated all plugins except: WPGraphQL, ACF Pro and WPGraphQL for ACF to avoid any possible plugin conflict.
3. Installed the Twenty Twenty Four theme to avoid any possible theme conflict.
4. The only modification to the original theme was to add a filter: `add_filter('use_block_editor_for_post_type', '__return_false', 10);` to remove gutenberg editor.
5. Create a new Field Group active in GraphQL with one normal text field and one flexible content field with a text field inside.
6. Added the Field Group to pages and posts.
7. Created a new post, updated content and then clicked on "Preview changes" button.
8. Used the query to view preview content but only the Post's title had a correct preview data. ACF field were still retrieving the published data. 

### PHP or JSON export of the ACF Field Group(s)

[acf-export-2024-01-16.json](https://github.com/wp-graphql/wpgraphql-acf/files/13956444/acf-export-2024-01-16.json)

### Additional context

_No response_

### WPGraphQL Version

1.19.0

### WPGraphQL For ACF Version

2.0.0 (download from WP plugin not github)

### ACF (Advanced Custom Fields) Version. Free or Pro?

6.2.5 PRO

### WordPress Version

6.4.2

### PHP Version

8.2.8

### Additional enviornment details

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have disabled ALL plugins except for WPGraphQL, WPGraphQL For ACF, ACF, etc.

- [X] Yes
- [ ] My issue is with compatibility with a specific WordPress plugin, and I have listed all my installed plugins (and version info) above.

### PR DESCRIPTION
**Pull request by jasonbahl**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This migrates some code from the previous version of WPGraphQL for ACF that got missed during the re-architecture. 

Does this close any currently open issues?
------------------------------------------
closes #154 

Any other comments?
-------------------
Because of the missing code that wasn't ported over during the re-architecture of the plugin, querying fields on posts while using `asPreview: true` were always returning null. 

Now, querying a preview will return the un-published revision data: 

### Published Post

The published post shows the latest values that were published.

CleanShot 2024-01-18 at 16 31 56

### Preview of the Post

If you click "preview" when editing a post (not Update or Publish), then query for it in WPGraphQL, you will see the un-published values, stored on the "revision" post.

CleanShot 2024-01-18 at 16 30 37

----

NOTE: The WordPress Block Editor (aka Gutenberg) has a bug that prevents Meta from being saved while using the block editor. This means that using ACF Field Groups as meta boxes on posts that also use the Block Editor beans the underlying data will not be stored on the preview and therefore cannot be resolved properly. 

Posts using the Block Editor will resolve the values from the parent (published) post instead when queried as a preview.

We have added a `graphql_debug` message to indicate this active Gutenberg issue: 

CleanShot 2024-01-18 at 16 36 25

### CODE REVIEW

**Comment by andremendonca03** [comment]
@jasonbahl Thanks for looking into it. 

Unfortunately I still can't preview acf fields. This time, however, the query doesn't retrieve published data anymore, just null.

1. When trying to query for a pages (post or cpt) preview, the ACF fields return null:

Screenshot 2024-01-19 at 1 09 47 pm

- Only the page's title returns the preview data as expected.
- One strange thing I notice is that when I click to preview the page, I'm still seeing the title "About Title" in WordPress default twenty-twenty-four theme (which in preview should display About Title Preview1 just as GraphQL does).

Screenshot 2024-01-19 at 1 13 55 pm

Just as a disclaimer: I'm not using any other plugin but WPGQL, ACF PRO and WPGQL-ACF with the changes from this PR. Default theme is twenty-twenty-four with a filter to remove guttenberg block editor.

**Comment by jasonbahl** [comment]
When you click preview are you right clicking to open in a new tab? If so that doesn't generate the preview data

**Comment by andremendonca03** [comment]
Wow... I really didn't know that. It worked!!

Honestly I don't know why I did't left-clicked it before ;)

Thanks immensely for that.

**Comment by jasonbahl** [comment]
Ya it's weird. But if you right click and open in a new tab it doesn't generate the nonce and so preview data isn't stored properly. 

I wonder if I can detect this somehow and output a debug message of some sort. 🤔

**Comment by andremendonca03** [comment]
Yeah that would definitely be helpful for others!

**Review by josephfusco — DISMISSED** [review]
I was able to pull this down to test.

I observed this working properly when the classic editor is active. When using the default block editor (Gutenberg), previews are not working as already called out in this PR. The additional debug notice was very helpful 🙌🏻

File: src/Registry.php [hunk 1]
```diff
@@ -46,8 +47,17 @@ public function __construct( TypeRegistry $type_registry = null ) {
...
+	 */
+	public function get_registered_fields() {
+		return $this->registered_fields;
+	}
+
 	/**
 	 * Get the TypeRegistry instance
+	 *
+	 * @return \WPGraphQL\Registry\TypeRegistry
 	 */
```
Comment (jasonbahl) [inline src/Registry.php:61]:
```suggestion
	/**
	 * Get the TypeRegistry instance
	 */
```

File: src/WPGraphQLAcf.php [hunk 2]
```diff
@@ -89,19 +97,101 @@ public function init_registry( TypeRegistry $type_registry ): void {
...
+	 *
+	 * NOTE: this currently only works if classic editor is not being used
+	 *
+	 * @param bool $should Whether to resolve using the parent object. Default true.
+	 * @param int $object_id The ID of the object to resolve meta for
+	 * @param string $meta_key The key for the meta to resolve
+	 * @param bool $single Whether a single value should be returned
+	 *
+	 * @return bool
+	 */
```
Comment (jasonbahl) [inline src/WPGraphQLAcf.php:152]:
```suggestion
	/**
	 * Add support for resolving ACF Fields when queried for asPreview
	 *
	 * NOTE: this currently only works if classic editor is not being used
	 *
	 * @param bool   $should Whether to resolve using the parent object. Default true.
	 * @param int    $object_id The ID of the object to resolve meta for
	 * @param string $meta_key The key for the meta to resolve
	 * @param bool   $single Whether a single value should be returned
	 */
```

**Pull request merged**